lib/SPIRVProducerPass.cpp

```diff
@@ -1568,18 +1521,40 @@ void SPIRVProducerPass::FindWorkgroupVars(Module &M) {
     if (LocalSpecIdInfoMap.count(spec_id))
       continue;
 
-    // We haven't seen this SpecId yet, so generate the LocalArgInfo for it.
-    LocalArgInfo info{nextID,     arg->getType()->getPointerElementType(),
-                      nextID + 1, nextID + 2,
-                      nextID + 3, spec_id};
-    LocalSpecIdInfoMap[spec_id] = info;
-    nextID += 4;
+    // Generate the spec constant.
+    SPIRVOperandVec Ops;
+    Ops << MkId(getSPIRVType(Type::getInt32Ty(Context))) << MkNum(1);
+    SPIRVID ArraySizeID = addSPIRVInst<kConstants>(spv::OpSpecConstant, Ops);
+
+    // Generate the array type.
+    Type *ElemTy = arg->getType()->getPointerElementType();
+    Ops.clear();
+    // The element type must have been created.
+    uint32_t elem_ty_id = getSPIRVType(ElemTy);
+    Ops << MkId(elem_ty_id) << MkId(ArraySizeID);
+
+    SPIRVID ArrayTypeID = addSPIRVInst<kTypes>(spv::OpTypeArray, Ops);
 
-    // Ensure the types necessary for this argument get generated.
-    Type *IdxTy = Type::getInt32Ty(M.getContext());
-    FindConstant(ConstantInt::get(IdxTy, 0));
-    FindType(IdxTy);
-    FindType(arg->getType());
+    Ops.clear();
+    Ops << MkNum(spv::StorageClassWorkgroup) << MkId(ArrayTypeID);
+    SPIRVID PtrArrayTypeID = addSPIRVInst<kTypes>(spv::OpTypePointer, Ops);
+
+    // Generate OpVariable.
+    //
+    // GIDOps[0] : Result Type ID
+    // GIDOps[1] : Storage Class
+    Ops.clear();
+    Ops << MkId(PtrArrayTypeID) << MkNum(spv::StorageClassWorkgroup);
+
+    SPIRVID VariableID = addSPIRVInst<kGlobalVariables>(spv::OpVariable, Ops);
+
+    Ops.clear();
+    Ops << MkId(ArraySizeID) << MkNum(spv::DecorationSpecId) << MkNum(spec_id);
+    addSPIRVInst<kAnnotations>(spv::OpDecorate, Ops);
+
+    LocalArgInfo info{VariableID,  ElemTy,         ArraySizeID,
+                      ArrayTypeID, PtrArrayTypeID, spec_id};
+    LocalSpecIdInfoMap[spec_id] = info;
```

Comment:
With SPIRV Sections the different components of Workgroup variables can be created all at once.